### ISSUE TITLE
TODO List

### ISSUE THREAD

**Issue by ldoshi** [ldoshi/rome-wasnt-built-in-a-day#1]
Edit the content of this ticket when claiming or adding items. It is ok for items to represent a broad or large piece of work; we can refine as we dig into it. 
* Claim by placing your initials to the left of the item and checking it. For non-trivial items, _**open a new issue**_ to facilitate discussion.
* Completed items should be moved to the TODONE section. 
* Ordering within a section does not bear any meaning.
* Priority is more in the direction sooner vs later, rather than importance to the project. We can add refinement along both axes later if it's deemed useful.

# TODO
### Higher Priority
- [ ] Understanding and improving NetworkQ architecture (including the use of Conv2D)
- [ ] Metrics for success for  {Graph of TD error training, td error validation, graph of rewards (optimal?)} in Tensorboard. Which samples to use?
- [ ] Scoring routine: Something like run n reset() and count how many build a bridge (and with what score)

### Lower Priority
- [ ] Debug network and determine base set of constant values in TrainingConfig to use while focusing on other pieces. Consider building a constant-choosing process (may need to revisit regularly)
- [x] [LD] Support generating gaps of varying with in the env (likely the env is fixed width, but the initial state changes?)
- [ ] Handling gaps of varying width
- [ ] Handling multiple gaps of varying width
- [ ] Add an action to declare a gap "un-bridgeable" and amend reward structure to reward giving up if a bridge is impossible (and penalize if done so incorrectly). TBD. If this is done, adjust environment generation to be less constrained.
- [ ] Network architecture evaluation and hyperparameter tuning tool to determine where things change, if we're over/under fitting.
- [ ] Network architecture -- develop simple baselines for comparison, including full random to verify we do better. 
- [ ] Handling gaps with varying heights on either side. This might already Just Work.
- [ ] Build on model save/load to enrich debugging/experimenting to go back to previous states
- [ ] Learn new "primitives" to increase building speed. This might be a primitive in terms of a "skill" -- how to build a bridge -- or a primitive in terms of a unit of blocks -- place an entire column or unit of fencing (constructed from individual bricks) as a single action. 
- [ ] Add a discriminator-type agent a la GANs to help train builder
- [ ] Aesthetic success criteria (bridge "quality"  / scoring function/rubric / GAN style?, Q functions for base reward system and other metrics like quality, Get intermediate penalties for failing against rubric, Loss is a combination of q functions)
- [ ] Consider exploration strategies aside from epsilon greedy ([link to start](https://medium.com/emergent-future/simple-reinforcement-learning-with-tensorflow-part-7-action-selection-strategies-for-exploration-d3a97b7cceaf))
- [ ] Consider Dueling DQN ([link to start](https://awjuliani.medium.com/simple-reinforcement-learning-with-tensorflow-part-4-deep-q-networks-and-beyond-8438a3e2b8df#.i2zpbmre8))
- [ ] Decide if we should limit world state to hide states during training for generalization check
- [ ] Build side-scrolling worlds into environment with fixed viewport and some mechanic to scroll left to right and configs to affect gap size (especially relative to viewport)
- [ ] Hierarchical approaches for managing new action primitives (can also have higher level (lower res view) and (zoomed in view for different components)
- [ ] A mechanism to checkpoint training and explore multiple branches of actions or "drill" certain scenarios (another notion of prioritizing training)

### Higher Priority
- [ ] Display and Update History Asynchronously During Training (#12)

### Lower Priority
- [ ] Add doc strings
- [ ] Model save/load support (so once doesn't need to train from scratch each time) -- backed by GCS free tier (for sharing)
- [ ] Make the training paradigm flexible and customizable. See below for notes.

# TODONE
## Functionality 
- [x] [LD] Incorporate TrainingHistory for debugging
- [x] [LD] Support generating varying heights in the env
- [x] [LD] Prioritized Experience Replay
- [x] [AT] Speed up (and generally improve) SumTree implementation 
- [x] [ALL] Determine if the NetworkQ style is sufficient the medium future re: 1 output per action. Decided ok for now.
- [x] [CX] Implement Double DQN

## Code
- [x] [AT] Switch to pytorch (#7)
- [x] [AT] Model save/load support (so once doesn't need to train from scratch each time) -- localhost (free with PyTorch Lightning, introduced in #7)

## Support
- [x] [LD] Write out project vision
- [x] [CX] update README

# Notes
**Make the training paradigm flexible and customizable**

Right now a fixed number of transitions (`inter_training_steps`) are generated and added to the replay buffer between consecutive (batch) training steps - the number of these to run, or the nature of memories to generate, should be customizable. This could be done by allowing the user to specify a stateful function that takes each output of the memory generator and returns True iff the program should stop and run a training batch before generating more memories. That said, there may be other/better designs for this.
